Port SelectVariants tool and dependent infrastructure.

cmnbroad opened the pull request.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/sam/ArtificialSAMUtils.java:
```diff
@@ -0,0 +1,70 @@
+package org.broadinstitute.hellbender.utils.sam;
```
Please delete this class -- it has been replaced in hellbender with `ArtificialReadUtils`

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/samples/Affection.java:
```diff
@@ -0,0 +1,22 @@
+package org.broadinstitute.hellbender.engine.samples;
```
I think this `samples` package would be better under `hellbender.utils` rather than `hellbender.engine`

cmnbroad commented:
Some comments/questions for the review:
- I ported a bunch of support files but need feedback on whether they're in the right locations.
-Somewhere I saw something that said GATK no longer supports .ped files ? If not, what should the replacement be in the tests require pedigree input?
-Is it a requirement to support Ploidy > 2 ? The current GATK tool, and thus the HB tool, do not
-I did not port the WalkerTestSpec.disableShadowVCF? Is that needed in Hellbender ?
-Are there other headers I should be applying to the output variant file ?

Command Line Arguments:
I did not port the GATK command line argument "-no_cmd_line_in_header". Should I ? And if not, should the command line args automatically be propagated to the output vcf file ? I didn't see GATK do this anywhere.
There was one test that used --variant:dbsnp on the command line but I couldn't find the code that processed that in GATK.
I replaced "-U LENIENT_VCF_PROCESSING" with "--lenient" (testFileWithoutInfoLineInHeaderWithOverride needs this to pass).
I replaced "-L" with --interval since HB seems to use -L for "lane" ?

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/StandardArgumentDefinitions.java:
```diff
@@ -11,12 +11,16 @@
     public static final String VARIANT_SHORT_NAME = "V";
     public static final String FEATURE_SHORT_NAME = "F";
+    public static final String LENIENT_SHORT_NAME = "LE";
+    public static final String INTERVAL_SHORT_NAME = "IN";
```
`-L` has been the standard interval argument in the GATK for a very long time -- we're really not at liberty to change this, given the amount of GATK tutorials, documentation, presentations, etc. that mention it.

lbergelson replied:
We should define constants here based on the ones in `IntervalArgumentCollection` though.

cmnbroad replied:
Makes sense. "-L" is already defined in this file as LANE_SHORT_NAME = "L" so I though it had been hijacked but I see it isn't referenced anywhere so I'll switch to that and reference IntervalArgumentCollection.

lbergelson replied:
Wierd, I wonder what lane short name was used by...

droazen commented:
Gave a bit of initial feedback -- will do a full review next week, and answer all of the questions you posted.

cmnbroad commented:
I added in some integration tests that use small files and get the coverage up to a reasonable level.

droazen commented:
@lbergelson let's co-review this one -- am a bit swamped in pull requests at the moment and could use some help

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/IntervalVariantFilter.java:
```diff
@@ -0,0 +1,30 @@
+
+    private final List<GenomeLoc> genomeLocs;
+
+    public IntervalVariantFilter(List<GenomeLoc> intervals) {
```
Prefer `Locatable` to `GenomeLoc` whenever possible.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/IntervalVariantFilter.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public boolean test( final VariantContext vc ) {
```
This is a prime use case for #559

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+    //The String is formatted "mom1+dad1=child1,mom2+dad2=child2,..."
+    public String getViolationFamiliesString() {
+        if (violationFamilies.isEmpty())
+            return "";
```
`{}` on the if statement

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+            return "";
+
+        final Iterator<String> it = violationFamilies.iterator();
+        String violationFams = it.next();
```
replace this block with a call to String.join()

lbergelson replied:
actually, this whole function can just be `return String.join(",", violationFamilies);` I think.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+     * @param abortOnSampleNotFound - Whether to stop execution if a family is passed but no relevant genotypes are found. If false, then the family is ignored.
+     */
+    public MendelianViolation(double minGenotypeQualityP, boolean abortOnSampleNotFound) {
+        minGenotypeQuality = minGenotypeQualityP;
```
Can this just delegate to the constructor below it?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+    }
+
+    /**
+     * Constructor
```
No need to have these say "constructor"

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+            Sample sample;
+            while(sampleIterator.hasNext()) {
+                sample = sampleIterator.next();
+                if(sample.getParents().size() > 0)
```
{}

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+
+        //Reset counts
+        nocall = 0;
+        lowQual = 0;
```
this block of resets should be extracted to it's own function and reused instead of repeated

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+        for (final Set<Sample> family : families.values()) {
+            final Iterator<Sample> sampleIterator = family.iterator();
+            Sample sample;
+            while(sampleIterator.hasNext()) {
```
could probably be rewritten more cleanly with a stream

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+        final Genotype gChild = vc.getGenotype(childId);
+
+        if (gMom == null || gDad == null || gChild == null){
+            if(abortOnSampleNotFound)
```
`{}`

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+            if(abortOnSampleNotFound)
+                throw new IllegalArgumentException(String.format("Variant %s:%d: Missing genotypes for family %s: mom=%s dad=%s family=%s",
+                        vc.getContig(), vc.getStart(), familyId, motherId, fatherId, childId));
+            else
```
`{}`

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+        return violationFamilies;
+    }
+
+    static final int[] mvOffsets = new int[] { 1,2,5,6,8,11,15,18,20,21,24,25 };
```
If you have any idea what theses numbers are a comment would be great.  I'm guessing you're equally mystified though.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+
+    //Count of HomRef/HomRef/HomRef trios
+    public int getRefRefRef() {
+        return inheritance.get(GenotypeType.HOM_REF).get(GenotypeType.HOM_REF).get(GenotypeType.HOM_REF);
```
I would argue that this class would benefit from at static import of GenotypeType despite our general stance against static imports.

lbergelson replied:
I might extract a method that does this get chanining.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+        double[] dadGL = vc.getGenotype(fatherId).getLikelihoods().getAsVector();
+        double[] childGL = vc.getGenotype(childId).getLikelihoods().getAsVector();
+        int offset = 0;
+        for (int oMom = 0; oMom < 3; oMom++) {
```
Could this 3 be named?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+     * @return the likelihood ratio for a mendelian violation
+     */
+    public double violationLikelihoodRatio(VariantContext vc, String motherId, String fatherId, String childId) {
+        double[] logLikAssignments = new double[27];
```
Where does this 27 come from?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/MendelianViolation.java:
```diff
@@ -0,0 +1,439 @@
+                }
+            }
+        }
+        double[] mvLiks = new double[12];
```
mysterious 12 and 15 should be named.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+ */
+public class PedReader {
+    private static Logger logger = LogManager.getLogger(PedReader.class);
+    final static private Set<String> CATAGORICAL_TRAIT_VALUES = new HashSet<String>(Arrays.asList("-9", "0", "1", "2"));
```
This should be declared using the named constants down below.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+ * @author Mark DePristo
+ * @since 2011
+ */
+public class PedReader {
```
final class

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+            final int paternalPos = missingFields.contains(MissingPedField.NO_PARENTS) ? -1 : samplePos + 1;
+            final int maternalPos = missingFields.contains(MissingPedField.NO_PARENTS) ? -1 : paternalPos + 1;
+            final int sexPos = missingFields.contains(MissingPedField.NO_SEX) ? -1 : Math.max(maternalPos, samplePos) + 1;
+            final int phenotypePos = missingFields.contains(MissingPedField.NO_PHENOTYPE) ? -1 : Math.max(sexPos, Math.max(maternalPos, samplePos)) + 1;
```
Ints.max(int ....) instead of nested max and `arrayMaxInt`

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+            // go through once and determine properties
+            int lineNo = 1;
+            boolean isQT = false;
+            final List<String[]> splits = new ArrayList<String[]>(lines.size());
```
lets generally try to replace all the specified types with `<>` when possible

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+            final List<String[]> splits = new ArrayList<String[]>(lines.size());
+            for (final String line : lines) {
+                if (line.startsWith(commentMarker)) {
+                    continue;
```
Isn't this check redundant?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+        return parse(new StringReader(source.replace(";", String.format("%n"))), missingFields, sampleDB);
+    }
+
+    public final List<Sample> parse(Reader reader, EnumSet<MissingPedField> missingFields, SampleDB sampleDB) {
```
This method is rather large could it be split up into several tighter methods?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+            }
+            return samples;
+
+        } catch (IOException e) {
```
This catch should probably move up so that it's just around that 1 line of IO code.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+        }
+    }
+
+    private final static String maybeMissing(final String string) {
```
final on a static method is redundant

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+            return string;
+    }
+
+    private final Sample maybeAddImplicitSample(SampleDB sampleDB, final String id, final String familyID, final Gender gender) {
```
likewise on a private method

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/PedReader.java:
```diff
@@ -0,0 +1,325 @@
+     *
+     * @param arg the actual engine arg, used for the UserException if there's an error
+     * @param tags a list of string tags that should be converted to the MissingPedField value
+     * @return
```
empty return tag, comment or delete

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/Sample.java:
```diff
@@ -0,0 +1,234 @@
+
+/**
+ *
+ */
```
Empty comment

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/Sample.java:
```diff
@@ -0,0 +1,234 @@
+/**
+ *
+ */
+public class Sample implements Comparable<Sample> { // implements java.io.Serializable {
```
What's with the commented out serializable?

lbergelson replied:
final

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/Sample.java:
```diff
@@ -0,0 +1,234 @@
+//    }
+//
+//    /**
+//     * Get one property
```
delete this commented out code

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/Sample.java:
```diff
@@ -0,0 +1,234 @@
+
+//    // -------------------------------------------------------------------------------------
+//    //
+//    // code for working with additional -- none standard -- properites
```
typo: properties

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/Sample.java:
```diff
@@ -0,0 +1,234 @@
+//    }
+
+    @Override
+    public int hashCode() {
```
It's weird that the hash code compares on different criteria than equals. There will be many hash collisions.

lbergelson replied:
Unless ID is unique... in which case equals seems redundant.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/SampleDB.java:
```diff
@@ -0,0 +1,317 @@
+import java.util.*;
+
+/**
+ * Simple databse for managing samples
```
typo: databse

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/SampleDB.java:
```diff
@@ -0,0 +1,317 @@
+    private final HashMap<String, Sample> samples = new HashMap<String, Sample>();
+
+    /**
+     * Constructor takes both a SAM header and sample files because the two must be integrated.
```
no it doesn't...

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/SampleDB.java:
```diff
@@ -0,0 +1,317 @@
+    }
+
+    public Set<Sample> getSamples() {
+        return new LinkedHashSet<>(samples.values());
```
This should probably just be a HashSet since `samples` is a `HashTable`

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/SampleDB.java:
```diff
@@ -0,0 +1,317 @@
+    /**
+     * Get a sample by its ID
+     * If an alias is passed in, return the main sample object
+     * @param id
```
floating param, remove or comment

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/SampleDB.java:
```diff
@@ -0,0 +1,317 @@
+
+    /**
+     *
+     * @param read
```
these useless tags should be made useful or removed.  Probably removed since there isn't really anything to say about them.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/SampleDB.java:
```diff
@@ -0,0 +1,317 @@
+     * @return - all the children that have both of their parents
+     * @deprecated - getTrios() replaces this function
+     */
+    @Deprecated
```
delete deprecated method

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/samples/SampleDB.java:
```diff
@@ -0,0 +1,317 @@
+     * .....................Child2
+     */
+    @Deprecated
+    public final Set<Sample> getChildrenWithParents(boolean triosOnly) {
```
delete

cmnbroad commented:
I still need to add a diffs commit for this for @droazen.

akiezun commented:
@droazen let me know when you're done - i'm happy to look at it too.

akiezun commented:
@droazen ping. let me know when you're done. happy to take this over

droazen commented:
Will get to this one tomorrow (was busy reviewing a large GATK3 pull request, hence the delay).

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/VariantWalker.java:
```diff
@@ -65,8 +65,13 @@ private void initializeDrivingVariants() {
+    void initializeIntervals() {
+        super.initializeIntervals();
+        if (hasIntervals()) {
+            // propagate any intervals for traversal; this needs to happen AFTER the intervals have been set up
```
Remove this fix the next time you rebase this branch -- a more general fix is about to be merged in https://github.com/broadinstitute/gatk/pull/976

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ExcludeIDsVariantFilter.java:
```diff
@@ -0,0 +1,27 @@
+ * Exclude variants with any of these IDs.
+ * Matching is done by case-sensitive exact match.
+ */
+public final class ExcludeIDsVariantFilter implements VariantFilter {
```
Name filters using predicate-style names that describe the variants that will be **kept** (rather than filtered out). See `ReadFilterLibrary` for naming examples.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/io/ListFileUtils.java:
```diff
@@ -0,0 +1,172 @@
+        // if they had been specified on the command line.
+        for (String value : values) {
+            File file = new File(value);
+            if (value.toLowerCase().endsWith(".list") && file.exists()) {
```
Extract an `isListFile()` method

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/io/ListFileUtils.java:
```diff
@@ -0,0 +1,172 @@
+     * @param exactMatch If true match filters exactly, otherwise use as both exact and regular expressions
+     * @return entries from values exluding any values matching filters
+     */
+    public static <T> Set<T> excludeMatching(Collection<T> values, StringConverter<T> converter, Collection<String> filters, boolean exactMatch) {
```
`includeMatching()` and `excludeMatching()` are almost identical apart from `add()` vs `remove()` -- recommend refactoring so that they call into a shared method.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/samples/MendelianViolation.java:
```diff
@@ -29,17 +29,18 @@
 
     /**
      * @param minGenotypeQualityP - the minimum phred scaled genotype quality score necessary to asses mendelian violation
      * @param abortOnSampleNotFound - Whether to stop execution if a family is passed but no relevant genotypes are found. If false, then the family is ignored.
```
Document the new third parameter

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/samples/MendelianViolation.java:
```diff
@@ -88,6 +92,29 @@ else if (minGenotypeQuality > 0 && (
     }
 
     /**
+     * @param sampleDB containing the families to be checked for Mendelian violations
```
Docs for `sampleIDs`

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/samples/Sample.java:
```diff
@@ -46,9 +58,18 @@ public Sex getSex() {
+
+    @Override
+    public int compareTo(final Sample sample) {
+        return ID.compareTo(sample.getID());
```
`compareTo()` and `equals()` are inconsistent (that is, `a.compareTo(b)` == 0 does not imply `a.equals(b)`) -- either add the standard disclaimer "This class has a natural ordering inconsistent with equals()` or fix this.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/samples/SampleDBBuilder.java:
```diff
@@ -0,0 +1,104 @@
+        //else {
+        //    final List<String> posTags = engine.getTags(engineArg).getPositionalTags();
+        //    return PedReader.parseMissingFieldTags(engineArg, posTags);
+        //}
```
Remove obsolete comment

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java:
```diff
@@ -267,6 +267,27 @@ public static File createTempFile(final String name, final String extension) {
+     * @param lines lines to write to the file.
+     * @return A list file in the temporary directory starting with tempFilePrefix, which will be deleted after the program exits.
+     */
+    public static File createTempListFile(final String tempFilePrefix, final String... lines) {
```
I don't think this method deserves to live in `BaseTest` (it's too specific) -- either move it to the test class that uses it or to a common utility class if multiple classes use it.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -114,6 +118,15 @@ public static int totalPloidy(final VariantContext vc, final int defaultPloidy)
+     * @return the genomeLoc
+     */
+    public static Locatable getLocation(final VariantContext vc) {
+        return new SimpleInterval(vc.getContig(), vc.getStart(), vc.getEnd());
```
Delete this method -- `VariantContext` is already a `Locatable`, so this conversion is totally unnecessary.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+    private static final double MIN_FRACTION_FILTERED_GENOTYPES_DEFAULT_VALUE = 0.0;
+
+    /**
+  n   * A site is considered discordant if there exists some sample in the variant track that has a non-reference
```
stray 'n' here in the margin

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+    public boolean lenientVCFProcesssing;
+
+    //@ArgumentCollection
+    //private IntervalArgumentCollection intervalList = new OptionalIntervalArgumentCollection();
```
Delete this commented-out argument collection (intervals come from the superclass)

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+    public void onTraversalStart() {
+        final FeatureInput<? extends Feature> featureInput = getDrivingVariantsFeatureInput();
+
+        final Map<String, VCFHeader> vcfRods = new HashMap<>();
```
Name this something like `vcfHeaders` -- "Rods" is an obsolete and confusing term

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+     */
+    @Override
+    public void onTraversalStart() {
+        final FeatureInput<? extends Feature> featureInput = getDrivingVariantsFeatureInput();
```
Use a better name here like `drivingVariantsFeatureInput`

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+        if (lenientVCFProcesssing) {
+            vcWriterBuilder = vcWriterBuilder.setOption(Options.ALLOW_MISSING_FIELDS_IN_HEADER);
+        }
+        vcfWriter = vcWriterBuilder.build();
```
Can you extract a utility method `createVCFWriter(File outFile, boolean indexOnTheFly, Options... additionalOptions)`, put it in `GATKVariantContextUtils`, and call it from here?

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+        final File refFile = referenceArguments.getReferenceFile();
+        final SAMSequenceDictionary samDictionary = this.getBestAvailableSequenceDictionary();
+        Set<VCFHeaderLine> actualLines;
+        if (refFile != null && samDictionary!= null) {
```
Call `hasReference()` here and branch based upon that. Could you also add a `getReferenceDictionary()` method to `GATKTool` and call that instead of `getBestAvailableSequenceDictionary()` when `hasReference()` returns true?

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+
+    @Override
+    public void apply(VariantContext vc, ReadsContext readsContext, ReferenceContext ref, FeatureContext featureContext) {
+
```
Where do you apply the `fullyDecode` option? In GATK3 we had:

            if (fullyDecode)
                vc = vc.fullyDecode(vcfRods.get(vc.getSource()), getToolkit().lenientVCFProcessing());

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+            return;
+        }
+
+        if (discordanceOnly && !isDiscordant(vc, featureContext.getValues(discordanceTrack, 0, 0))) {
```
Call the one-arg overload of `getValues()` here: `getValues(final FeatureInput<T> featureDescriptor)`

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+            return;
+        }
+
+        if (concordanceOnly && !isConcordant(vc, featureContext.getValues(concordanceTrack, 0, 0))) {
```
Here too -- call the one-arg overload of `getValues()`: `getValues(final FeatureInput<T> featureDescriptor)`

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+        else if (refPath.contains("hg19"))
+            assembly = "hg19";
+        return assembly;
+    }
```
I think we should delete this terrible method and just not include the assembly header line, since the reference is optional in the hellbender version of this tool anyway. @akiezun do you agree?

akiezun replied:
agreed. this is horrid

vdauwera replied:
Ack, that's horrible and should definitely die. But what's the functional implication of getting rid of this? Do we lose information in the header, or does this affect our ability to valid the reference dictionary? What was that info used for previously, do we know?

cmnbroad replied:
In GATK3, this code actually lives in VCFUtils.smartMergeHeaders, so it affects more than just SelectVariants. Its only used there to add "assembly=something" tags to the contig header lines - I don't know if there are any tools that try to consume it.

droazen replied:
I don't know of anything in the GATK that programmatically consumes these header lines -- I think they were just for the benefit of human readers of the header. Since we can't add this header line if there's no reference, and since we expect to pretty much always run `SelectVariants` without a reference in GATK4, we don't really lose anything by remove this functionality from the GATK4 version of `SelectVariants`.

vdauwera replied:
OK, then feel free to just kill it.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -0,0 +1,1244 @@
+        // @TODO: this TODO was in GATK - is this still the case ?
+        if (fullyDecode) {
+            return; // TODO -- annotations are broken with fully decoded data
+        }
```
I don't think you are ever actually fully decoding the records, even if `fullyDecode` is true, unless I've missed something (see my comment above).

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariantsIntegrationTest.java:
```diff
@@ -0,0 +1,751 @@
+
+    private static final String b36KGReference = "/humgen/1kg/reference/human_b36_both.fasta";
+    private static final String b37KGReference = "/humgen/1kg/reference/human_g1k_v37.fasta";
+    private static final String GATKDataLocation = "/humgen/gsa-hpprojects/GATK/data/";
```
We need to completely remove the dependency on files in the Broad filesystem (and remove the `requiresRemoteAccess` group), even if it means changing the test data quite a bit. Could you perhaps refactor to use slices of `src/test/resources/large/dbsnp_138.b37.1.1-65M.vcf` as your input, and not use a reference at all in the tests (since it's optional in the hellbender version of the tool)?

droazen replied:
(or if there are only a few large vcfs, and they're not too huge, we could check them into `src/test/resources/large` using `git lfs`)

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariantsIntegrationTest.java:
```diff
@@ -0,0 +1,751 @@
+    // but now fails with: UserException$LexicographicallySortedSequenceDictionary: A USER ERROR has occurred:
+    // Lexicographically sorted human genome sequence detected in variants. From debugging it appears that the
+    // variants file the source of the error message, but it still fails the same way even after running the
+    // picard SortVcf tool on the .vcf using the b37KGReference reference dictionary.
```
Can you edit the sequence dictionary manually? You just need to ensure that chromosome 10 comes after chromosomes 1 and 2.

droazen commented:
Review complete -- looks good overall! A few requested changes to the tool itself, and one major issue in the `SelectVariantsIntegrationTest` with reliance on files in the Broad filesystem -- I've proposed a way to eliminate that (mainly by running `SelectVariants` without a reference). Back to @cmnbroad. @akiezun should also have a look if he has time.

droazen commented:
To answer a few of your earlier questions that may not have been answered:

-The hellbender equivalent of `--variant:dbsnp vcf` is `--variant dbsnp:vcf`. Ie., `--argument_name logical_name:feature_file` instead of `--argument_name:logical_name feature_file`

cmnbroad commented:
Removed large file dependencies, all tests are enabled, and all code changes requested are done. The previous version of this branch is at cn_select_variants_bu.

droazen commented on src/test/resources/bams.out:
```diff
@@ -0,0 +1,105 @@
+./large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam
```
This shouldn't be here?

droazen commented:
:+1: Looks good -- merging this one at last!
